### ISSUE TITLE
Eliminate or catch potential panic in ProcessProposal

### ISSUE THREAD

**Issue by evan-forbes** [celestiaorg/celestia-app#595]
When checking the data root during ProcessProposal, https://github.com/celestiaorg/celestia-app/blob/f5a07435f5969bb558b72f4189611db9ff1eefad/app/process_proposal.go#L119

its possible for a malicious block proposer to gossip a block with messages that are out of order, which would make some nodes in the network panic. It is unclear how long this proposal block would be gossiped, as any nodes receive it would panic, but we should still prevent this. 

We should make the block validity rule that messages are sorted by namespace more explicit and avoid panicking if this rule is not followed.

**Comment by rootulp** [comment]
> a block with messages that are out of order

Based on [Message Block Layout](https://celestiaorg.github.io/celestia-specs/latest/rationale/message_block_layout.html), I assume "out of order" is with respect to namespace IDs. But we have [a test](https://github.com/celestiaorg/celestia-app/blob/d52d89ef5507c94122b7dd6bfb27cc0b108ced27/app/test/process_proposal_test.go#L38-L57) that has a message list with random namespace IDs per message so it seems like ordering of namespace IDs doesn't result in a rejected block.

In other words, it is not clear to me why this test passes: https://github.com/celestiaorg/celestia-app/compare/main...rootulp:celestia-app:rp/message-order-test

**Comment by evan-forbes** [comment]
great point! yes, so _currently_ it won't panic because the messages are automatically implicitly (:fearful:) sorted. when computing the shares, we call a function that encodes the messages into shares https://github.com/celestiaorg/celestia-core/blob/60b268964c495926b294e69b488f10a7293f5ace/types/block.go#L1084
which calls a function to sort the messages for this exact reason, to avoid an accidental panic. This however will not be the case in the future, when the messages will not be able to be sorted (due to their order being important to following the non-interactive default rules) 
https://github.com/celestiaorg/celestia-core/blob/60b268964c495926b294e69b488f10a7293f5ace/types/block.go#L1166

also, it is a block validity rule, so we just need to be super explicit about it which we currently are not. As another side note, by sorting implicitly, we're also currently relying on this implementation to be deterministic!!! It appears to be deterministic, but if there many messages of the same namespace, that would be something else that we have to check (we don't have worry about this in the future as we won't be able to sort)

instead of sorting, we just need to check if the shares are sorted, and reject the block if not.

**Comment by evan-forbes** [comment]
reference https://github.com/celestiaorg/celestia-core/pull/546

**Comment by rootulp** [comment]
If I understand correctly, we want to remove the implicit sorting [here](https://github.com/celestiaorg/celestia-core/blob/60b268964c495926b294e69b488f10a7293f5ace/types/block.go#L1166) and add a conditional in `ProcessProposal` (somewhere before `ComputerShares` is called so [here](https://github.com/rootulp/celestia-app/blob/2534b4dec532ede5aa819f7628376454f981fa38/app/process_proposal.go#L109)) that checks if the messages are sorted and if not, log an error and REJECT the block.

If this is correct, how can we make this change atomically given they span two repos? I don't think it's possible so I think we need to:
1. add the conditional in celestia-app
2. remove the sort from celestia-core
3. cut a celestia-core release
4. upgrade to that version in celestia-app

**Comment by evan-forbes** [comment]
I'm actually changing the part in core (and moving to the app!) with the non-interactive defaults, so I think we should either wait until that is finished or just do the app portion and create an issue to make sure we remove the sorting.

**Comment by evan-forbes** [comment]
its also worth noting that we're currently only sorting messages, so a block proposer could still put things in namespaces that reserved and possibly cause a panic. When checking that things are sorted, we need to remember to check the entire set of shares
